buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java

```diff
@@ -0,0 +1,354 @@
+            new PooledByteBufAllocator(numArenas, numArenas, 8192, 1);
+
+        CountDownLatch tcache0 = createNewThreadCache(allocator);
+        CountDownLatch tcache1 = createNewThreadCache(allocator);
+        assertEquals(2, allocator.numThreadLocalCaches());
+        destroyThreadCache(tcache1);
+        assertEquals(1, allocator.numThreadLocalCaches());
+
+        CountDownLatch tcache2 = createNewThreadCache(allocator);
+        assertEquals(2, allocator.numThreadLocalCaches());
+
+        destroyThreadCache(tcache0);
+        assertEquals(1, allocator.numThreadLocalCaches());
+
+        destroyThreadCache(tcache2);
+        assertEquals(0, allocator.numThreadLocalCaches());
+    }
+
+    private static void destroyThreadCache(CountDownLatch tcache) {
+        tcache.countDown();
+        LockSupport.parkNanos(MILLISECONDS.toNanos(100));
+    }
+
+    private static CountDownLatch createNewThreadCache(final PooledByteBufAllocator allocator)
+            throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final CountDownLatch cacheLatch = new CountDownLatch(1);
+        Thread t = new FastThreadLocalThread(new Runnable() {
+
+            @Override
+            public void run() {
+                ByteBuf buf = allocator.newHeapBuffer(1024, 1024);
+
+                // Countdown the latch after we allocated a buffer. At this point the cache must exists.
+                cacheLatch.countDown();
+
+                buf.writeZero(buf.capacity());
+
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    throw new IllegalStateException(e);
+                }
+
+                buf.release();
+
+                FastThreadLocal.removeAll();
+            }
+        });
+        t.start();
+
+        // Wait until we allocated a buffer and so be sure the thread was started and the cache exists.
+        cacheLatch.await();
+
+        return latch;
+    }
+
+    @Test
+    public void testConcurrentUsage() throws Throwable {
+        long runningTime = MILLISECONDS.toNanos(SystemPropertyUtil.getLong(
+                "io.netty.buffer.PooledByteBufAllocatorTest.testConcurrentUsageTime", 15000));
+
+        // We use no caches and only one arena to maximize the chance of hitting the race-condition we
+        // had before.
+        ByteBufAllocator allocator = new PooledByteBufAllocator(true, 1, 1, 8192, 11, 0, 0, 0);
+        List<AllocationThread> threads = new ArrayList<AllocationThread>();
+        try {
+            for (int i = 0; i < 512; i++) {
+                AllocationThread thread = new AllocationThread(allocator);
+                thread.start();
+                threads.add(thread);
+            }
+
+            long start = System.nanoTime();
+            while (!isExpired(start, runningTime)) {
+                checkForErrors(threads);
+                Thread.sleep(100);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)